src/main/java/bio/terra/service/tabulardata/google/BigQueryProjectProvider.java

```diff
@@ -0,0 +1,21 @@
+package bio.terra.service.tabulardata.google;
+
+import org.springframework.stereotype.Component;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * Given a projectId, returns a {@link BigQueryProject} object while caching
+ */
+@Component
+public class BigQueryProjectProvider implements Function<String, BigQueryProject> {
+    private static final ConcurrentHashMap<String, BigQueryProject> BQ_PROJECT_CACHE = new ConcurrentHashMap<>();
+
+    @Override
+    public BigQueryProject apply(String projectId) {
+        BQ_PROJECT_CACHE.computeIfAbsent(projectId, BigQueryProject::new);
+        return BQ_PROJECT_CACHE.get(projectId);
+    }
+
+}
```

Comment:
Is there a benefit to subclassing `Function`? I would expect to see:
```suggestion
public class BigQueryProjectProvider {
    private static final ConcurrentHashMap<String, BigQueryProject> BQ_PROJECT_CACHE = new ConcurrentHashMap<>();

    @Override
    public BigQueryProject create(String projectId) {
        BQ_PROJECT_CACHE.computeIfAbsent(projectId, BigQueryProject::new);
        return BQ_PROJECT_CACHE.get(projectId);
    }
}
```

Reply:
I went back and forth.  I like the idea of having explicitly be a function.  It communicates the intent of the class (e.g. it's a functional interface).

Reply:
But it doesn't need to inherit from `Function` to be used in a functional way, any class with a single method will work this way.